SPEC.md

```diff
@@ -0,0 +1,195 @@
+   2. The container runtime reads the JSON file.
+   3. The container runtime validates that the device is described in the JSON file.
+   4. The container runtime pulls the container image.
+   5. The container runtime generates an OCI specification.
+   6. The container runtime transforms the OCI specification according to the instructions in the JSON file
+
+
+[cni]: https://github.com/containernetworking/cni
+[oci]: https://github.com/opencontainers/runtime-spec
+
+## General considerations
+
+- The device configuration is in JSON format and can easily be stored in a file.
+- The device configuration includes mandatory fields such as "name" and "version".
+- Fields in CDI structures are required unless specifically marked optional.
+
+For the purposes of this proposal, we define two terms very specifically:
+- _container_ can be considered synonymous with a [Linux _network namespace_][namespaces].
+  What unit this corresponds to depends on a particular container runtime implementation. 
+- _device_ which will refer to an actual hardware device.
+
+The key words "must", "must not", "required", "shall", "shall not", "should", "should not", "recommended", "may" and "optonal" are used as specified in [RFC 2119][rfc-2119].
+
+[rfc-2119]: https://www.ietf.org/rfc/rfc2119.txt
+[namespaces]: http://man7.org/linux/man-pages/man7/namespaces.7.html
+
+## CDI JSON Specification
+
+### JSON Definition
+
+```
+{
+    "cdiVersion": "0.1.0",
+    "kind": "<name>",
+    "container-runtime": ["<container-runtime-name>"], (optional)
+
+    "cdiDevices": [
+        {
+            "name": "<name>",
+            "nameShort": ["<short-name>", "<short-name>"], (optional)
+
+            // Same as the below containerSpec field.
+            // This field should only be applied to the Container's OCI spec
+            // if that specific device is requested.
+            "containerSpec": { ... }
+        }
+    ],
+
+    // This field should be applied to the Container's OCI spec if any of the
+    // devices defined above are requested on the CLI
+    "containerSpec": [
+        {
+            "devices": [ (optional)
+                {
+                    "hostPath": "<path>",
+                    "containerPath": "<path>",
+
+                    // Cgroups permissions of the device, candidates are one or more of
+                    // * r - allows container to read from the specified device.
+                    // * w - allows container to write to the specified device.
+                    // * m - allows container to create device files that do not yet exist.
+                    "permissions": "<permissions>" (optional)
+                }
+            ]
+            "mounts": [ (optional)
+                {
+                    "hostPath": "<source>",
+                    "containerPath": "<destination>",
+                    "options": "<OCI Mount Options>", (optional)
+                }
+            ],
+            "hooks": [ (optional)
```

Comment:
I do not see a stage keyword for when the hook is going to be activated, are we assuming that the hook will always be fired and read the first argument supplied by the runtime to know in which stage the runtime currently is?

Reply:
This means that this interface (runtime providing the stage in arg1) has to be stable and always valid/available.

Reply:
What about adding an OCI "activationStage" for the hooks?

Reply:
I'm not sure I understand can you give an example of what a hook might want to do?

Reply:
I do not know if all runtimes have this kind of configuration as CRI-O has, but in CRI-O the hook description you can state in which stage you want a specific hook to be activated and other activations. 

The question is how will the runtime know in which stage to activate/run the hook?

Reply:
Another thing I am curious about is which stages of a runtime do we want to tackle? Should we make it possible to use "any" stage (createRuntime, createContainer, startContainer, poststart, poststop) or are we abstracting and having something like: beforeCommand, afterCommand (where command is either CMD or ENTRYPOINT)

Reply:
So I think the "stage" you mention is actually called "hookName" in this spec.
For example, you could have the following spec:
```
    "hooks": [
      {hookName: "create-container", "path": "/bin/vendor-hook" },
    ]
```

Reply:
I think one of the biggest discussion that needs to happen is if we want to have hookName only allow OCI hook names or if it's ok for it to have hooks that are not understood (e.g: for lxc)